### PR TITLE
Species Infrastructure

### PR DESCRIPTION
**Pull request by JustinPrivitera**
Some changes to blueprint index and the addition of a `to_silo` utility for the full species set case + a test.
All changes are to facilitate efforts down the road in VisIt's Blueprint reader and Conduit's Silo I/O capability.

### CODE REVIEW

File: src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp [hunk 1]
```diff
@@ -1425,6 +1426,263 @@ to_multi_buffer_by_material(const conduit::Node &src_matset,
...
+    {
+        CONDUIT_ERROR("blueprint::mesh::specset::to_silo passed specset node"
+                      " must be a valid specset tree.");
+    }
+
+    if(!matset.dtype().is_object() )
+    {
+        CONDUIT_ERROR("blueprint::mesh::specset::to_silo passed matset node"
+                      " must be a valid matset tree.");
+        // TODO well, not entirely true...
```
Comment (JustinPrivitera) [inline src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp:1449]: I want to give users the option to just pass the silo representation of the matset directly into this function if they already have one. Or they can pass a regular matset. Either way, this function will be smart and create the silo matset if it has to. So this error message is somewhat dishonest... should I change it? Also, is that even a good idea?

File: src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp [hunk 2]
```diff
@@ -1425,6 +1426,263 @@ to_multi_buffer_by_material(const conduit::Node &src_matset,
...
+
+    // we have to go in order by zones as they appear
+
+    // first we need number of zones
+    const int nzones = silo_matset["matlist"].dtype().number_of_elements();
+
+    // TODO
+    // I may wish to go through and check if the material is even in the zone
+    // to avoid writing unneeded data
+    // that could be expensive though
```
Comment (JustinPrivitera) [inline src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp:1532]: This is an optimization to explore later, if at all. It won't affect correctness of the result.

File: src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp [hunk 3]
```diff
@@ -1425,6 +1426,263 @@ to_multi_buffer_by_material(const conduit::Node &src_matset,
...
+                                                decltype(nmatspec)::value_type(1));
+
+        // we save the final index for this zone
+        return outer_index + local_index;
+
+        // TODO return a zero here if the material in the zone contains only 1 species.
+        // This is a further optimization. It doesn't matter for now since we are
+        // treating our output like we have all species and all materials in all zones.
+        // I think the point of this optimization is to also leave stuff out of the 
+        // species_mf array. If I want to do this then I should explore that as well.
```
Comment (JustinPrivitera) [inline src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp:1605]: Again, also an optimization to explore down the road.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp [hunk 4]
```diff
@@ -2364,15 +2364,21 @@ mesh::generate_index_for_single_domain(const Node &mesh,
...
-            }
+                const Node &matset_val = matset_vals_itr.next();
+                const std::string matname = matset_vals_itr.name();
 
+                NodeConstIterator specs_itr = matset_val.children();
+                while(specs_itr.has_next())
+                {
+                    specs_itr.next();
+                    const std::string specname = specs_itr.name();
+                    idx_specset["species"][matname][specname];
```
Comment (JustinPrivitera) [inline src/libs/blueprint/conduit_blueprint_mesh.cpp:2379]: Are there other things I should do for specset index verify now that I have changed this node?

File: src/tests/blueprint/t_blueprint_mesh_matset_xforms.cpp [hunk 5]
```diff
@@ -475,3 +475,35 @@ TEST(conduit_blueprint_mesh_matset_xforms, mesh_util_matset_full_to_sparse_by_el
...
+TEST(conduit_blueprint_mesh_matset_xforms, mesh_util_to_silo_misc)
+{
+    Node mesh;
+    blueprint::mesh::examples::misc("specsets", 4, 4, 1, mesh);
+    const Node &matset = mesh["matsets/mesh"];
+    const Node &specset = mesh["specsets/mesh"];
+
+    Node silo_rep, silo_rep_matset, info;
+    blueprint::mesh::specset::to_silo(specset, matset, silo_rep);
+    blueprint::mesh::matset::to_silo(matset, silo_rep_matset);
```
Comment (JustinPrivitera) [inline src/tests/blueprint/t_blueprint_mesh_matset_xforms.cpp]: TODO @JustinPrivitera why is this here

File: src/libs/blueprint/conduit_blueprint_mesh.cpp (hunk 4 above)
Comment (JustinPrivitera) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp:2379]: Cyrus says no

File: src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp (hunk 1 above)
Comment (JustinPrivitera) [reply src/libs/blueprint/conduit_blueprint_mesh_matset_xforms.cpp:1449]: Yes good idea but change message

**Comment by JustinPrivitera** [comment]
@cyrush this is ready for real now :)

**Review by cyrush — APPROVED** [review]
Looks good!
